In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from category_encoders import OneHotEncoder

# Load the dataset
# url = "../sourcedata/cleaned_data/6_weather_parameters.csv"
# df = pd.read_csv(url)
# df

# Read the file and check it out
df = pd.read_csv(r'/Users/jberndt/Desktop/2-2-opencampus/X-opencampus_ML/2_DS_and_ML/6-homework_week06/6_weather_parameters.csv')
df.info()
df

Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6999 entries, 0 to 6998
Data columns (total 19 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Datum                6999 non-null   object 
 1   Bewoelkung           6999 non-null   float64
 2   Temperatur           6999 non-null   float64
 3   Windgeschwindigkeit  6999 non-null   float64
 4   Wettercode           6999 non-null   float64
 5   KielerWoche          6999 non-null   f

Datum  Bewoelkung  Temperatur  Windgeschwindigkeit  Wettercode  \
0     2013-07-01         6.0     17.8375                 15.0        20.0   
1     2013-07-01         6.0     17.8375                 15.0        20.0   
2     2013-07-01         6.0     17.8375                 15.0        20.0   
3     2013-07-01         6.0     17.8375                 15.0        20.0   
4     2013-07-01         6.0     17.8375                 15.0        20.0   
...          ...         ...         ...                  ...         ...   
6994  2018-07-30         6.0     27.7875                 10.0        21.0   
6995  2018-07-30         6.0     27.7875                 10.0        21.0   
6996  2018-07-30         6.0     27.7875                 10.0        21.0   
6997  2018-07-30         6.0     27.7875                 10.0        21.0   
6998  2018-07-30         6.0     27.7875                 10.0        21.0   

      KielerWoche  Warengruppe      Umsatz  Feiertage  Ferientage  \
0             0.0          1.0  148.828353        0.0         1.0   
1             0.0          2.0  535.856285        0.0         1.0   
2             0.0          3.0  201.198426        0.0         1.0   
3             0.0          4.0   65.890169        0.0         1.0   
4             0.0          5.0  317.475875        0.0         1.0   
...           ...          ...         ...        ...         ...   
6994          0.0          1.0  148.541340        0.0         1.0   
6995          0.0          2.0  616.358562        0.0         1.0   
6996          0.0          3.0  348.770346        0.0         1.0   
6997          0.0          4.0   71.822977        0.0         1.0   
6998          0.0          5.0  324.965348        0.0         1.0   

      Wetterklasse  Niederschlag  year  month  week  weekday  day_month  \
0                6           0.3  2013      7    27        0          1   
1                6           0.3  2013      7    27        0          1   
2                6           0.3  2013      7    27        0          1   
3                6           0.3  2013      7    27        0          1   
4                6           0.3  2013      7    27        0          1   
...            ...           ...   ...    ...   ...      ...        ...   
6994             6           0.0  2018      7    31        0         30   
6995             6           0.0  2018      7    31        0         30   
6996             6           0.0  2018      7    31        0         30   
6997             6           0.0  2018      7    31        0         30   
6998             6           0.0  2018      7    31        0         30   

     season_str  season  
0        summer       2  
1        summer       2  
2        summer       2  
3        summer       2  
4        summer       2  
...         ...     ...  
6994     summer       2  
6995     summer       2  
6996     summer       2  
6997     summer       2  
6998     summer       2  

[6999 rows x 19 columns]

In [2]:
# Define the column that we want to predict
X = df.drop(['Umsatz', 'Datum', 'Wettercode', 'year', 'month', 'week', 'season_str'], axis=1)  # Exclude the columns that shall not be included
y = df['Umsatz']

# Identify numerical columns and categorical columns
numerical_cols = X.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_cols = X.select_dtypes(include=['object', 'bool']).columns.tolist()

# Add 'Wetterklasse' to categorical_cols
categorical_cols.append('Wetterklasse')

# Exclude weather code from numerical columns
int_or_float_cols = ['Bewoelkung', 'KielerWoche', 'Warengruppe', 'Feiertage', 'Ferientage', 'day_month']
numerical_cols = [col for col in numerical_cols if col not in int_or_float_cols]

# Initialize the encoder
encoder = OneHotEncoder(cols=categorical_cols)

# Fit and transform the categorical columns
X_encoded = encoder.fit_transform(X)

# Combine the original numerical columns with the encoded categorical columns
X_final = pd.concat([X[numerical_cols], X_encoded], axis=1)

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_final, y, test_size=0.3, random_state=42)

# Create and fit the linear regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
predictions = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, predictions)
print(f"Mean Squared Error: {mse}")

Mean Squared Error: 19653.813325191844


/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:757: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:595: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype):
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:604: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:757: FutureWarning: is_sparse is deprecated and will be rem

It seems that the model is not performing well, as indicated by the high Mean Squared Error (MSE) value of 19653.813325191844. This error measure quantifies the average difference between the true and predicted values, so a large MSE indicates that the model's predictions are significantly different from the actual values. There are several possible reasons for a high MSE value, including:

1. **Inadequate model**: The chosen model (Linear Regression) might not be the best model for the given data. You could try other regression models like Ridge Regression, Lasso Regression, Elastic Net, or even machine learning algorithms like Random Forest Regressor or Gradient Boosting Regressor.

2. **Feature engineering**: The current feature set might not be sufficient to capture the relationship between the input and output variables. You could consider adding new features, transforming the existing features, or selecting different features using techniques like Principal Component Analysis (PCA) or Recursive Feature Elimination (RFE).

3. **Data quality**: The data might contain errors, missing values, or outliers that could negatively impact the model's performance. You could consider cleaning the data, removing outliers, or handling missing values using appropriate techniques.

4. **Model hyperparameters**: The current model hyperparameters might not be optimal. You could experiment with different hyperparameters, such as regularization strength or learning rate, to improve the model's performance.

To improve the model's performance, you could try the following steps:

1. **Experiment with different models**: Try using other regression models or machine learning algorithms and compare their performance using various evaluation metrics.

2. **Feature engineering**: Investigate the relationship between the input features and the output variable and engineer new features or transform the existing features to improve the model's performance.

3. **Data cleaning**: Clean the data by removing errors, handling missing values, and removing outliers.

4. **Model hyperparameter tuning**: Experiment with different hyperparameter values to find the optimal settings for the chosen model.

5. **Ensemble methods**: Use ensemble methods like Random Forest Regressor, Gradient Boosting Regressor, or bagging to combine multiple models and improve the overall performance.

To improve the model performance and reduce the Mean Squared Error (MSE), you can consider the following strategies:

1. Feature Engineering:
- Explore additional features that may have predictive power for the target variable 'Umsatz'.
- Consider creating new features based on domain knowledge or interactions between existing features.
- Check for multicollinearity among the features and remove redundant or highly correlated features.

2. Data Preprocessing:
- Standardize or normalize numerical features to ensure they are on a similar scale.
- Handle missing values in the dataset using appropriate techniques such as imputation or removal.
- Encode categorical variables effectively, considering different encoding methods like target encoding or frequency encoding.

3. Model Selection and Tuning:
- Experiment with different regression models such as Ridge Regression, Lasso Regression, or Decision Trees to see if they perform better than Linear Regression.
- Perform hyperparameter tuning using techniques like GridSearchCV or RandomizedSearchCV to find the optimal parameters for the chosen model.
- Consider ensemble methods like Random Forest or Gradient Boosting for potentially improved performance.

4. Cross-Validation:
- Implement cross-validation techniques like k-fold cross-validation to assess the model's generalization performance and reduce overfitting.

5. Regularization:
- Apply regularization techniques like L1 (Lasso) or L2 (Ridge) regularization to prevent overfitting and improve model generalization.

By incorporating these strategies and fine-tuning your model, you can work towards improving its performance and reducing the Mean Squared Error. Remember to evaluate the model's performance on both the training and testing datasets to ensure it generalizes well to unseen data.

In [3]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from sklearn.feature_selection import SelectKBest, f_regression

# Your DataFrame
X = df.drop(['Datum', 'season_str'], axis=1)  # Exclude the columns that are not numeric
y = df['Umsatz']

# Correlation matrix
corr = X.corr().abs()
print("Correlation matrix:\n", corr)

# Selecting features based on correlation score
selector = SelectKBest(score_func=f_regression, k=5)
X_new = selector.fit_transform(X, y)
X_selected = pd.DataFrame(X.iloc[:, selector.get_support(indices=True)].T, columns=X.columns[selector.get_support(indices=True)])
print("\nFeatures selected using SelectKBest with R2 score:")
print(X_selected)

# Linear Regression
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Selecting the features using the indices
selected_columns = X.columns[selector.get_support(indices=True)]
X_train_selected = X_train[selected_columns]
X_test_selected = X_test[selected_columns]

linear_regression = LinearRegression()
linear_regression.fit(X_train_selected, y_train)

# Predictions
y_pred = linear_regression.predict(X_test_selected)

# Evaluation
print("\nLinear Regression Results with Selected Features:")
print("R2 score:", r2_score(y_test, y_pred))
print("Mean squared error:", mean_squared_error(y_test, y_pred))
print("Mean absolute error:", mean_absolute_error(y_test, y_pred))


Correlation matrix:
                      Bewoelkung  Temperatur  Windgeschwindigkeit  Wettercode  \
Bewoelkung             1.000000    0.293302             0.003829    0.524281   
Temperatur             0.293302    1.000000             0.057264    0.085217   
Windgeschwindigkeit    0.003829    0.057264             1.000000    0.120354   
Wettercode             0.524281    0.085217             0.120354    1.000000   
KielerWoche            0.004290    0.166036             0.016537    0.067041   
Warengruppe            0.020495    0.042958             0.009703    0.009581   
Umsatz                 0.063891    0.224926             0.027708    0.009880   
Feiertage              0.031130    0.035565             0.000431    0.005797   
Ferientage             0.043719    0.198472             0.043579    0.035105   
Wetterklasse           0.538271    0.074773             0.198274    0.741259   
Niederschlag           0.226376    0.043528             0.139073    0.438680   
year               

/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:757: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:595: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype):
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:604: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:757: FutureWarning: is_sparse is deprecated and will be rem

Based on the provided results, the linear regression model with the selected features has a relatively low R-squared score, indicating that only a small portion of the variance in the target variable (Umsatz) can be explained by the selected features. Additionally, the mean squared error and mean absolute error are relatively high, suggesting that the model has limited predictive power.

To improve the model, you can consider the following steps:

1. Check for data preprocessing and feature engineering: Ensure that the data is clean and preprocessed properly. Feature engineering techniques like normalization, scaling, or feature transformations can help improve model performance.

2. Try different models: You can experiment with different regression models or machine learning algorithms to identify if there is a better model for your dataset. For instance, you can try Ridge Regression, Lasso Regression, or Elastic Net.

3. Feature selection: You may also want to try different feature selection techniques like Lasso Regression with `alpha=0` (L1 regularization), or Elastic Net with a mix of L1 and L2 regularization. These methods perform both regression and feature selection in a single step, which can improve the model's performance.

4. Cross-validation: You can use k-fold cross-validation to evaluate the model's performance and generalizability. This can help you identify if the model is overfitting or underfitting the data.

5. Model tuning: You can try tuning the model's hyperparameters using techniques like Grid Search or Random Search to find optimal values.

In [4]:
# Based on the given code and the correlation matrix, it seems that the "Temperatur" and "KielerWoche" columns have relatively high correlation coefficients with the target variable "Umsatz." You might want to consider keeping these two features in your model instead of using SelectKBest with R2 score, which appears to be selecting irrelevant features.


import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error

# Your DataFrame
X = df[['Temperatur', 'KielerWoche']]  # Select the desired columns
y = df['Umsatz']

# Linear Regression
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Selecting the features using the indices
linear_regression = LinearRegression()
linear_regression.fit(X_train, y_train)

# Predictions
y_pred = linear_regression.predict(X_test)

# Evaluation
print("\nLinear Regression Results with Selected Features:")
print("R2 score:", r2_score(y_test, y_pred))
print("Mean squared error:", mean_squared_error(y_test, y_pred))
print("Mean absolute error:", mean_absolute_error(y_test, y_pred))


# This updated code selects only the 'Temperatur' and 'KielerWoche' columns for the linear regression model. These two columns have relatively high correlation with the target variable "Umsatz" as per the correlation matrix.


Linear Regression Results with Selected Features:
R2 score: 0.04972141045782785
Mean squared error: 18844.98408692643
Mean absolute error: 105.3104643418714


/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:757: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:595: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype):
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:604: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:757: FutureWarning: is_sparse is deprecated and will be rem

The R2 score, mean squared error, and mean absolute error are all measures of how well the linear regression model fits the data. A lower R2 score indicates that the model explains a smaller percentage of the variance in the data, while a lower mean squared error and mean absolute error indicate better prediction accuracy.

In this case, the updated model with 'Temperatur' and 'KielerWoche' as features has a lower R2 score (0.0497) compared to the original R2 score (0.0764). This means that the model explains a smaller percentage of the variance in the data with these two features.

However, the mean squared error and mean absolute error have also decreased slightly compared to the original values. This suggests that the model with 'Temperatur' and 'KielerWoche' as features still provides some improvement in prediction accuracy compared to the full model with all features.

Ultimately, the choice of which model to use depends on the specific goals and constraints of your analysis. If explaining a larger percentage of the variance in the data is a priority, you might consider keeping more features in the model. If prediction accuracy is more important, the model with 'Temperatur' and 'KielerWoche' as features could be a reasonable choice.

It's also worth noting that there are other techniques, such as stepwise regression and Lasso/Ridge regression, that can be used to select and fit features in a more systematic way. These methods might provide better results in terms of model performance and interpretability.

In [5]:
# Import Ridge Regression
from sklearn.linear_model import Ridge

# Define the column that we want to predict
X = df.drop(['Umsatz', 'Datum', 'Wettercode', 'year', 'month', 'week', 'season_str'], axis=1)  # Exclude the columns that shall not be included
y = df['Umsatz']

# Identify numerical columns and categorical columns
numerical_cols = X.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_cols = X.select_dtypes(include=['object', 'bool']).columns.tolist()

# Add 'Wetterklasse' to categorical_cols
categorical_cols.append('Wetterklasse')

# Exclude weather code from numerical columns
int_or_float_cols = ['Bewoelkung', 'KielerWoche', 'Warengruppe', 'Feiertage', 'Ferientage', 'day_month']
numerical_cols = [col for col in numerical_cols if col not in int_or_float_cols]

# Initialize the encoder
encoder = OneHotEncoder(cols=categorical_cols)

# Fit and transform the categorical columns
X_encoded = encoder.fit_transform(X)

# Combine the original numerical columns with the encoded categorical columns
X_final = pd.concat([X[numerical_cols], X_encoded], axis=1)

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_final, y, test_size=0.3, random_state=42)

# Implement Feature Selection using Ridge Regression
# Initialize the Ridge Regression model
ridge_reg = Ridge(alpha=1)

# Fit the model
ridge_reg.fit(X_train, y_train)

# Get intercept and coefficients separately
intercept = pd.DataFrame(ridge_reg.intercept_, index=X_train.columns, columns=['Intercept'])
coefs = pd.DataFrame(ridge_reg.coef_, index=X_train.columns, columns=['Coefficient'])

# Convert columns to numeric type
coefs = coefs.astype(float)

# Sort the columns based on their absolute coefficients
sorted_indices = sorted(coefs.columns, key=lambda x: abs(coefs[x][0]), reverse=True)[:10]

# Print the column names of the top 10 features
print("Top 10 Features:")
for index in sorted_indices:
    print(X_train.columns[index])
    # print(X_train.columns[int(index)])

# Create and fit the Ridge Regression model using the selected features
X_selected = X_train[X_train.columns[sorted_indices]]  # Select the top 19 features
y_selected = y_train
ridge_reg_selected = Ridge(alpha=1)
ridge_reg_selected.fit(X_selected, y_selected)

# Make predictions on the test set
predictions = ridge_reg_selected.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, predictions)
print(f"Mean Squared Error: {mse}")

/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:757: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if not hasattr(array, "sparse") and array.dtypes.apply(is_sparse).any():
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:595: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype):
/opt/anaconda3/envs/tensor_course/lib/python3.10/site-packages/sklearn/utils/validation.py:604: FutureWarning: is_sparse is deprecated and will be removed in a future version. Check `isinstance(dtype, pd.SparseDtype)` instead.
  if is_sparse(pd_dtype) or not is_extension_array_dtype(pd_dtype):
/var/folders/df/036b7scd3f90j03plj3rnhrxnr0csc/T/ipykernel_45804/2198933905.py:46: FutureWarning: Series.__getitem__ treating keys as positions is dep

Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
Top 10 Features:


IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices